magneticod/magneticod/dht.py

```diff
@@ -30,107 +31,95 @@
 
 
 class SybilNode:
-    def __init__(self, address: typing.Tuple[str, int]):
+    def __init__(self, address: typing.Tuple[str, int], complete_info_hashes, max_metadata_size):
         self.__true_id = self.__random_bytes(20)
 
-        self.__socket = socket.socket(type=socket.SOCK_DGRAM)
-        self.__socket.bind(address)
-        self.__socket.setblocking(False)
+        self.__address = address
 
-        self.__incoming_buffer = array.array("B", (0 for _ in range(65536)))
-        self.__outgoing_queue = collections.deque()
-
-        self.__routing_table = {}  # type: typing.Dict[NodeID, NodeAddress]
+        self._routing_table = {}  # type: typing.Dict[NodeID, NodeAddress]
 
         self.__token_secret = self.__random_bytes(4)
         # Maximum number of neighbours (this is a THRESHOLD where, once reached, the search for new neighbours will
         # stop; but until then, the total number of neighbours might exceed the threshold).
         self.__n_max_neighbours = 2000
+        self.__tasks = {}  # type: typing.Dict[dht.InfoHash, asyncio.Future]
+        self._complete_info_hashes = complete_info_hashes
+        self.__max_metadata_size = max_metadata_size
+        self._metadata_q = asyncio.Queue()
+        self._is_paused = False
+        self._tick_task = None
 
         logging.info("SybilNode %s on %s initialized!", self.__true_id.hex().upper(), address)
 
-    @staticmethod
-    def when_peer_found(info_hash: InfoHash, peer_addr: PeerAddress) -> None:
-        raise NotImplementedError()
+    async def launch(self, loop):
+        self._loop = loop
+        await loop.create_datagram_endpoint(lambda: self, local_addr=self.__address)
 
-    def on_tick(self) -> None:
-        self.__bootstrap()
-        self.__make_neighbours()
-        self.__routing_table.clear()
+    def connection_made(self, transport):
+        self._tick_task = self._loop.create_task(self.on_tick())
+        self._transport = transport
 
-    def on_receivable(self) -> None:
-        buffer = self.__incoming_buffer
-        while True:
-            try:
-                _, addr = self.__socket.recvfrom_into(buffer, 65536)
-                data = buffer.tobytes()
-            except BlockingIOError:
-                break
-            except ConnectionResetError:
-                continue
-            except ConnectionRefusedError:
-                continue
-
-            # Ignore nodes that uses port 0 (assholes).
-            if addr[1] == 0:
-                continue
+    def connection_lost(self, exc):
+        self._is_paused = True
```

Comment:
I don't remember if this being called or not. I may have just added it when I added connection_made. You could try logging or putting a `pdb.set_trace()` to see if it's ever called. I believe it's called when `close` is called on the transport. https://github.com/python/cpython/blob/master/Lib/asyncio/selector_events.py#L621